### DISCUSSION TITLE
Using Nearley

### DISCUSSION THREAD

**Discussion by nexensys** [ScratchAddons/ScratchAddons#2102] [Help] [answered]
Would it interfere with any part of the extension if I used [nearley](https://nearley.js.org) for parsing in my keyboard editor addon?

**Comment by cobaltt7** [comment, marked as answer]
No. But using it would be a little more complicated.

We can't run any scripts from an external website due to CWS guidelines. So you have to download the source into the `libraries` folder in this repository. Update `license-info.json` and `CREDITS.md` (nearly is on the MIT license). Add a `libraries` key to your `addon.json`. It should be an array of library names. 

When you want to use it in your addon, run `await addon.tab.loadScript(addon.self.lib + <filename of the library source>)`.

**Reply by nexensys** [reply]
I am thinking of trying to modify the code so we can do something like

```js
const parser = Nearley.load(`${addon.self.dir}/path/to/file.ne`);
parser.parse("code here");
```
so that it can be done dynamically and without node.js.

**Reply by cobaltt7** [reply]
You can do it dynamically. But you will need to use node to [compile your grammar](https://nearley.js.org/docs/getting-started#:~:text=Step,n%22%20%5D%5D%5D). After you do that, upload `grammar.js` to your addon's folder.

---

As I said earlier, download the source into the libraries folder in this repository. Name it `nearley.js`. [You can find the nearley source code here](https://cdn.jsdelivr.net/npm/nearley@2.20.1/lib/nearley.min.js).

---

Update `license-info.json` with
```
"nearley": "MIT"
```
Add this to `CREDITS.md`:
```
- [nearley 2.20.1](https://cdn.jsdelivr.net/npm/nearley@2.20.1/lib/nearley.min.js) (MIT)
```
 Add a `libraries` key to your `addon.json`.:
```
"libraries": ["nearley"]
```

---

When you want to use it in your addon, run `await addon.tab.loadScript(addon.self.lib + "nearley.js")`.
I forgot to mention earlier that you will also have to run `await addon.tab.loadScript(addon.self.dir + "grammar.js")`.

**Reply by nexensys** [reply]
Okay, but I am actually going to use [this](https://omrelli.ug/nearley-playground/) to compile. I do not currently have access to node (for reasons of my own). I will add the script to a library then.

**Reply by cobaltt7** [reply]
Ok, that shouldn't cause any problems.

**Reply by nexensys** [reply]
Will I need to clean up grammar.js or should I add it to the Prettier ignore list?

**Reply by cobaltt7** [reply]
Either should be fine.

(You could also surround your code with `// prettier-ignore-start` and `// prettier-ignore-end` to ignore it.)

**Reply by nexensys** [reply]
Okay, thanks!

**Reply by nexensys** [reply]
Actually, turns out it won't work anyway... I need to be able to dynamically add variables, sprite names, costumes, and custom blocks.
